semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/linking/filecompat/FileLinker.scala

```diff
@@ -0,0 +1,64 @@
+package io.shiftleft.semanticcpg.passes.linking.filecompat
+
+import gremlin.scala._
+import io.shiftleft.codepropertygraph.Cpg
+import io.shiftleft.codepropertygraph.generated.{EdgeTypes, NodeTypes, nodes}
+import io.shiftleft.semanticcpg.passes.linking.linker.Linker
+import io.shiftleft.passes.{CpgPass, DiffGraph}
+
+import scala.jdk.CollectionConverters._
+import scala.collection.mutable
+
+/**
+  * Links NAMESPACE_BLOCKs, TYPE_DECLs and METHODs to their FILEs via
+  * SOURCE_FILE edges.
+  *
+  * This pass should come after FILENAME properties have been recovered,
+  * that is, after FileNameCompat.
+  */
+class FileLinker(cpg: Cpg) extends CpgPass(cpg) {
+  override def run(): Iterator[DiffGraph] = {
+    val dstGraph = DiffGraph.newBuilder
+
+    val linker = new Linker(cpg)
+
+    val originalFileNameToNode = mutable.Map.empty[String, nodes.StoredNode]
+    val newFileNameToNode = mutable.Map.empty[String, nodes.NewFile]
+
+    var maxFileOrder = -1
+
+    cpg.graph.graph.vertices().asScala.foreach {
```

Comment:
Do `cpg.file...` instead of iterating over all vertices this will only iterate over the FILE nodes.